lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/impl/MultiObserverImpl.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.reef.wake.impl;
+
+import org.apache.reef.wake.MultiObserver;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The MultiObserverImpl class uses reflection to discover which onNext()
+ * event processing methods are defined and then map events to then.
+ * @param <TSubCls> The subclass derived from MultiObserverImpl.
+ */
+public abstract class MultiObserverImpl<TSubCls> implements MultiObserver {
+  private static final Logger LOG = Logger.getLogger(MultiObserverImpl.class.getName());
+  private final Map<String, Method> methodMap = new HashMap<>();
+  private boolean initialized = false;
+
+  /**
+   * Use reflection to discover all of the event processing methods in TSubCls
+   * and setup a means to direct calls from the generic event onNext method defined
+   * in the MultiObserver interface to specific concrete event onNext methods.
+   */
+  private void initialize() {
+    // Get all of the onNext event processing methods.
+    final Class<?> cls = this.getClass();
+    final Method[] methods = cls.getMethods();
+
+    // Iterate across the methods and build a hash map of class names to reflection methods.
+    for (int idx = 0; idx < methods.length; ++idx)  {
+      if (methods[idx].getName() == "onNext" && methods[idx].getDeclaringClass().equals(this.getClass())) {
+        // This is an onNext method defined in TSubCls
+        final Class<?>[] types = methods[idx].getParameterTypes();
+        if (types.length == 2 && types[0].getSimpleName() == "long") {
+          methodMap.put(types[1].getName(), methods[idx]);
+        }
+      }
+    }
+  }
+
+  /**
+   * Called when an event is received that does not have an onNext method definition
+   * in TSubCls. Override in TSubClas to handle the error.
+   * @param event A reference to an object which is an event not handled by TSubCls.
+   */
+  public void unimplemented(final long identifier, final Object event) {
```

Comment:
* Method is not used by any external classes and can be declared `private`;
   * Consider making method generic using `<TEvent>` instead of using `Object` parameter